src/transformers/models/sam/image_processing_sam.py

```diff
@@ -0,0 +1,658 @@
+    points_x = np.tile(points_one_side[None, :], (n_per_side, 1))
+    points_y = np.tile(points_one_side[:, None], (1, n_per_side))
+    points = np.stack([points_x, points_y], axis=-1).reshape(-1, 2)
+    return points
+
+
+def _build_all_layer_point_grids(
+    points_per_crop: int = None, n_layers: int = None, scale_per_layer: int = None
+) -> List[np.ndarray]:
+    """Generates point grids for all crop layers."""
+    points_by_layer = []
+    for i in range(n_layers + 1):
+        n_points = int(points_per_crop / (scale_per_layer**i))
+        points_by_layer.append(_build_point_grid(n_points))
+    return points_by_layer
+
+
+def _normalize_coordinates(target_size, coords: np.ndarray, original_size, is_bounding_box=False) -> np.ndarray:
+    """
+    Expects a numpy array of length 2 in the final dimension. Requires the original image size in (H, W) format.
+    """
+    old_h, old_w = original_size
+    new_h, new_w = _get_preprocess_shape(original_size[0], original_size[1], target_size)
+    coords = deepcopy(coords).astype(float)
+
+    if is_bounding_box:
+        # reshape to .reshape(-1, 2, 2)
+        coords = coords.reshape(-1, 2, 2)
+
+    coords[..., 0] = coords[..., 0] * (new_w / old_w)
+    coords[..., 1] = coords[..., 1] * (new_h / old_h)
+
+    if is_bounding_box:
+        # reshape back to .reshape(-1, 4)
+        coords = coords.reshape(-1, 4)
+
+    return coords
+
+
+def _generate_crop_boxes(
+    image,
+    target_size,
+    n_layers: int = 0,
+    overlap_ratio: float = 512 / 1500,
+    points_per_crop: Optional[int] = 32,
+    scale_per_layer: Optional[List[int]] = 1,
+) -> Tuple[List[List[int]], List[int]]:
+    """
+    Generates a list of crop boxes of different sizes. Each layer has (2**i)**2 boxes for the ith layer.
+    """
+    if isinstance(image, list):
+        raise ValueError("Only one image is allowed for crop generation.")
+    image = to_numpy_array(image)
+    original_size = image.shape[:2]
+
+    points_grid = _build_all_layer_point_grids(
+        points_per_crop=points_per_crop, n_layers=n_layers, scale_per_layer=scale_per_layer
+    )
+    crop_boxes, layer_idxs = [], []
+    im_h, im_w = original_size
+    short_side = min(im_h, im_w)
+
+    # Original image
+    crop_boxes.append([0, 0, im_w, im_h])
+    layer_idxs.append(0)
+
+    def crop_len(orig_len, n_crops, overlap):
+        return int(math.ceil((overlap * (n_crops - 1) + orig_len) / n_crops))
+
+    for i_layer in range(n_layers):
+        n_crops_per_side = 2 ** (i_layer + 1)
+        overlap = int(overlap_ratio * short_side * (2 / n_crops_per_side))
+
+        crop_w = crop_len(im_w, n_crops_per_side, overlap)
+        crop_h = crop_len(im_h, n_crops_per_side, overlap)
+
+        crop_box_x0 = [int((crop_w - overlap) * i) for i in range(n_crops_per_side)]
+        crop_box_y0 = [int((crop_h - overlap) * i) for i in range(n_crops_per_side)]
+
+        # Crops in XYWH format
```

Comment:
It would be more helpful for future contributors and the maintenance if the names were more explicit:
```suggestion
        # Crops in XYWH format (left, top, width, height)
```
then all x -> left, y->top, w->width and h->height

w and h are explicit enough but x and y could be the coordinates of the center or any other corner.

Reply:
These functions and all the functions below have been removed as they are specific to the AMG pipeline. We will address that in https://github.com/huggingface/transformers/pull/22840